powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/BaseProvider.java

```diff
@@ -0,0 +1,195 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package software.amazon.lambda.powertools.parameters;
+
+import software.amazon.awssdk.annotations.NotThreadSafe;
+import software.amazon.lambda.powertools.parameters.exception.TransformationException;
+import software.amazon.lambda.powertools.parameters.internal.DataStore;
+import software.amazon.lambda.powertools.parameters.transform.BasicTransformer;
+import software.amazon.lambda.powertools.parameters.transform.Transformer;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+/**
+ * Base class for all parameter providers.
+ */
+@NotThreadSafe
+public abstract class BaseProvider {
+
+    static final int DEFAULT_MAX_AGE_SECS = 5;
+
+    private final DataStore store = new DataStore();
+    private int defaultMaxAge = DEFAULT_MAX_AGE_SECS;
+    private int maxAge = defaultMaxAge;
+    private Class<? extends Transformer> transformerClass;
+
+    /**
+     * Retrieve the parameter value from the underlying parameter store.<br />
+     * Abstract: Implement this method in a child class of {@link BaseProvider}
+     *
+     * @param key key of the parameter
+     * @return the value of the parameter identified by the key
+     */
+    abstract String getValue(String key);
+
+    /**
+     * (Optional) Set the default max age (in seconds) for the cache of all parameters. Override the default 5 seconds.<br/>
+     * If for some parameters, you need to set a different maxAge, use {@link #withMaxAge(int)}.
+     *
+     * @param maxAgeinSec Maximum time (in seconds) to cache the parameter, before calling the underlying parameter store.
+     * @return the provider itself in order to chain calls (eg. <code>provider.defaultMaxAge(10).get("key")</code>).
+     */
+    public <T extends BaseProvider> T defaultMaxAge(int maxAgeinSec) {
+        this.defaultMaxAge = maxAgeinSec;
+        this.maxAge = maxAgeinSec;
+        return (T) this;
+    }
+
+    /**
+     * (Optional) Builder method to call before {@link #get(String)} or {@link #get(String, Class)}
+     * to set cache max age (in seconds) for the parameter to get.<br/><br/>
+     * The max age is reset to default (either 5 or a custom value set with {@link #defaultMaxAge}) after each get,
+     * so you need to use this method for each parameter to cache with non-default max age.<br/><br/>
+     *
+     * <b>Not Thread Safe</b>: calling this method simultaneously by several threads
+     * can lead to unwanted cache time for some parameters.<br/>
+     *
+     * @param maxAgeinSec Maximum time (in seconds) to cache the parameter, before calling the underlying parameter store.
+     * @return the provider itself in order to chain calls (eg. <code>provider.withMaxAge(10).get("key")</code>).
+     */
+    public <T extends BaseProvider> T withMaxAge(int maxAgeinSec) {
```

Comment:
Would it be better to support withMaxAge(ChronoUnit unit, int amount)?

I can imagine most parameters would have quite long durations. Over a minute, often over an hour.

Reply:
Yeah makes sense to me.

Reply:
Makes sense, just a bit more long to type...

Reply:
Yeah we could probably have both ? one with seconds like it is now and other with time unit